### PR TITLE
6264 test mobile phone resolution

### PR DESCRIPTION
**Pull request by ngaruko**
# Description

Created a new test suite for 'mobile emulation' and copied some passing  and easy to refactor tests to the new test suite.
Created a grunt task (`e2e-mobile`) for the test
Added a new `mobile-conf` file with `mobileEmulation` option
Updated development setup documentation

medic/cht-core#6264

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/mobile/login /login.specs.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+const auth = require('../../auth')();
```
Comment (garethbowen) [inline tests/mobile/login /login.specs.js:1]: This path has a space at the end of `/login /` - can you move it to remove the space?

File: tests/page-objects/forms/generic-form.po.js [hunk 2]
```diff
@@ -1,4 +1,11 @@
 const helper = require('../../helper');
+const waitForLoaderToDisappear = () => {
+  try {
+    helper.waitElementToDisappear(by.css('.loader'));
```
Comment (garethbowen) [inline tests/page-objects/forms/generic-form.po.js:4]: We use `.loader` everywhere so this is probably not specific enough - can you provide a more specific selector before the `.loader` so it's more reliable?

File: tests/page-objects/login/login.po.js [hunk 3]
```diff
@@ -30,6 +30,7 @@ module.exports = {
     changeLocale(locale);
     getLoginButton().click();
     browser.waitForAngular();
+    browser.sleep(200);//seems to be failing here...maybe local browser slow
```
Comment (garethbowen) [inline tests/page-objects/login/login.po.js:33]: We'll want to get to the bottom of this one...

**Review by garethbowen — CHANGES_REQUESTED** [review]
Nice! Green build!

I've left a few comments inline... additionally:

- There are quite a few unrelated changes like formatting and so on. I suspect your editor automatically formatted files. Can you view the diff and revert all the changes that aren't required for this to succeed? This might require changing the settings in your editor.
- There's some commented out code which should be removed too.
- It looks like you copied the browser resolution tests over to the mobile suite. I think they should be removed from the desktop suite so we're only running a single e2e test. Ideally you would use `git mv` so the git history would be maintained.
- This change adds 3 more builds which will make the CI build much longer because we're limited to only 5 concurrent builders. I think instead the travis build should first run the desktop resolution tests, and then run the mobile resolution tests in the same build job. That way it only has to be deployed once and it only needs a single builder.

File: tests/mobile-conf.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+const constants = require('./constants');
```
Comment (newtewt) [inline tests/mobile-conf.js:1]: This is also defined in grunt file and seems to be redefined here. I think we can get rid of the extra conf file and use the existing one. 

Define the new mobile suite in the `tests/conf.js` and pass the arguments in the grunt file. Which I believe you're already doing.

```
      'e2e-tests-mobile': {
        options: {
          configFile: 'tests/mobile-conf.js',
          args: {
            suite: 'mobile'
          },
          capabilities: {
            chromeOptions: {
              mobileEmulation: { 'deviceName': 'Nexus 5' }
            }
          }
        }
      }
```

File: tests/mobile/common/common.specs.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+const commonElements = require('../../page-objects/common/common.po.js');
+const utils = require('../../utils');
+
+describe('Navigation tests : ', () => {
+  beforeEach(utils.beforeEach);
+
+  it('should open Messages tab', () => {
+    commonElements.goToMessages();
+    expect(commonElements.isAt('message-list'));
+    //expect(browser.getCurrentUrl()).toMatch(utils.getBaseUrl() + 'messages/');
```
Comment (newtewt) [inline tests/mobile/common/common.specs.js:11]: By sending to a mobile size does it break this expect?

**Review by newtewt — COMMENTED** [review]
Awesome! Are you finding that the navigation and flow of tests is so different that we need to have a second set of spec files for mobile? I would hope we could abstract navigation away in the page object (we should be able set a isMobile check) and the test scenarios remain the same regardless of browser.

File: tests/page-objects/login/login.po.js (hunk 3 above)
Comment (ngaruko) [reply tests/page-objects/login/login.po.js]: removed - it seems stable enough on travis...something slow on my machine -

File: tests/page-objects/forms/generic-form.po.js (hunk 2 above)
Comment (ngaruko) [reply tests/page-objects/forms/generic-form.po.js:4]: removed - it seems stable enough on travis...something slow on my machine -

File: tests/mobile/login /login.specs.js (hunk 1 above)
Comment (ngaruko) [reply tests/mobile/login /login.specs.js]: Fixed

**Review by garethbowen — CHANGES_REQUESTED** [review]
There are lots of formatting or whitespace changes which is making it difficult to review. Can you please clean this up and then apply for review again?

**Comment by ngaruko** [comment]
> There are lots of formatting or whitespace changes which is making it difficult to review. Can you please clean this up and then apply for review again?

Fixed @garethbowen - Apologies... Looked quite messy. I need to fix settings with my editor

**Review by garethbowen — COMMENTED** [review]
Awesome! Looks great.

I'm a little concerned about the code that starts and stops API based on the name of the suite - it'll work well for travis, but if you just want to run one suite locally then it won't work as expected.

I had a look at [the documentation](https://www.npmjs.com/package/grunt-protractor-runner) and it looks like we can specify an array of suites, eg:

```json
args: {
  suite: ['e2e', 'mobile'],
```

This would mean we could have three grunt tasks - one for running e2e, one for mobile, and one for both. It's not clear if the afterLaunch and onPrepare code runs before all suites or not but it might be worth investigating.

An alternative approach would be to move the API starting and stopping code to grunt somehow, but that's a bigger job unfortunately...

**Review by garethbowen — APPROVED** [review]
Awesome work! Thanks. No AT required - merge when ready.

**Pull request merged**